pkg/nfd-topology-updater/nfd-topology-updater.go

```diff
@@ -290,3 +305,38 @@ func updateAttributes(lhs *v1alpha2.AttributeList, rhs v1alpha2.AttributeList) {
 		updateAttribute(lhs, attr)
 	}
 }
+
+func getKubeletConfigFunc(uri, apiAuthTokenFile string) (func() (*kubeletconfigv1beta1.KubeletConfiguration, error), error) {
```

Comment:
Ideally, I'd prefer having a goroutine waiting for changes in the kubeletconfig and re-reading it if needed and only if needed, avoding constant and costly polling. IIRC we use already `fsnotify` in NFD so this would mean no new deps. Have you explored this option? how hard would it be to implement?

Reply:
I feel like if we want to support both endpoints and not to overcomplicate the code. We can spin up another goroutine that will check `kubelet` config every 5mins(or make it configurable) to reduce costly poling and `patch` only `TopologyPolices` or attributes field?

Reply:
Fair point. I'm hesitant to add kubelet config read in this path, this is why I mentioned a different goroutine.
Let me review the code again and think a bit more.

Reply:
I feel like spinning up another goroutine which will be updating NRT object can lead to conflicts when updating same resource it would be best to have one func that's responsible for managing NRT. We can also consider such implementation:
```
-       policy, scope, err := w.detectTopologyPolicyAndScope()
-       if err != nil {
-               return fmt.Errorf("failed to detect TopologyManager's policy and scope: %w", err)
-       }
-
-       tmAttributes := createTopologyAttributes(policy, scope)
-       deprecatedTopologyPolicies := []string{string(topologypolicy.DetectTopologyPolicy(policy, scope))}
-
        nrt, err := cli.TopologyV1alpha2().NodeResourceTopologies().Get(context.TODO(), w.nodeName, metav1.GetOptions{})
        if errors.IsNotFound(err) {
+               policy, scope, err := w.detectTopologyPolicyAndScope()
+               if err != nil {
+                       return fmt.Errorf("failed to detect TopologyManager's policy and scope: %w", err)
+               }
+
+               tmAttributes := createTopologyAttributes(policy, scope)
+               deprecatedTopologyPolicies := []string{string(topologypolicy.DetectTopologyPolicy(policy, scope))}
+
                nrtNew := v1alpha2.NodeResourceTopology{
                        ObjectMeta: metav1.ObjectMeta{
                                Name: w.nodeName,
@@ -237,15 +239,28 @@ func (w *nfdTopologyUpdater) updateNodeResourceTopology(zoneInfo v1alpha2.ZoneLi
        nrtMutated := nrt.DeepCopy()
        nrtMutated.Zones = zoneInfo

-       attributes := append(scanResponse.Attributes, tmAttributes...)
-       updateAttributes(&nrtMutated.Attributes, attributes)
+       if w.lastUpdateTimestamp.IsZero() || time.Since(w.lastUpdateTimestamp).Minutes() > float64(5) {
+               policy, scope, err := w.detectTopologyPolicyAndScope()
+               if err != nil {
+                       return fmt.Errorf("failed to detect TopologyManager's policy and scope: %w", err)
+               }

-       nrtMutated.TopologyPolicies = deprecatedTopologyPolicies
+               tmAttributes := createTopologyAttributes(policy, scope)
+               deprecatedTopologyPolicies := []string{string(topologypolicy.DetectTopologyPolicy(policy, scope))}
+
+               attributes := append(scanResponse.Attributes, tmAttributes...)
+               updateAttributes(&nrtMutated.Attributes, attributes)
+
+               nrtMutated.TopologyPolicies = deprecatedTopologyPolicies
+       }

        nrtUpdated, err := cli.TopologyV1alpha2().NodeResourceTopologies().Update(context.TODO(), nrtMutated, metav1.UpdateOptions{})
        if err != nil {
                return fmt.Errorf("failed to update NodeResourceTopology: %w", err)
        }
+
+       w.lastUpdateTimestamp = time.Now()
+
```

or move reading `scope` and `policy` into seperate goroutine and keep them in shared state somewhere?

Reply:
@ffromani to reduce number of times we need to read Kubelet config file and minimize complexity. I have modified the code only to reread Kubelet config when `Interval based` event is triggered, which default to 1 min(somewhat low). Should we introduce another ticker? 
I feel like we should implement a better solution to this error but I am unsure if we need to do it in this PR.

Reply:
thanks @PiotrProkop my main issue was finding time to review, I'll have another pass ASAP. I tend to agree without a some redesign (which is better served on a separate PR) we can't do much better than the current approach.